module/src/hooks/clientside-validation/formValidationHook.tsx

```diff
@@ -0,0 +1,132 @@
+ * the input has been touched.
+ * @param dataBinder the form state
+ * @param inputValidators  the form validators
+ */
+export function calculateFormValidity<T>(formState: T, inputValidators: InputValidator[]): boolean {
+  for (const inputValidator of inputValidators) {
+    for (const validator of inputValidator.validators) {
+      if (!validator.validate(valueByKeyChain(formState, inputValidator.key), formState)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+/**
+ * Converts a key chain to dot notation.
+ * @param key the key chain to convert
+ * @returns the key chain in dot notation
+ */
+function convertKeyChainToDotNotation(keyChain: KeyChain): string {
+  return keyChain.filter((key) => !!key).join('.');
+}
+
+/**
+ * Runs validation on the given input. This sets the error message to the first piece of validation that fails.
+ * Otherwise this removes all errors if the input is valid
+ * @param input input to test validity against
+ * @param state the state of the form
+ * @param formProp formProp
+ */
+function runValidationOnInput<TData>(input: InputValidator, state: TData, formProp: (...keyChain: KeyChain) => BindingTools<TData>) {
+  const value = valueByKeyChain(state, input.key);
+
+  formProp(...input.key).clearValidationErrors();
+  for (const validator of input.validators) {
+    if (!validator.validate(value, state)) {
+      formProp(...input.key).addValidationError(validator.message);
+      break;
+    }
+  }
+};
+
+/**
+ * Hook to enable real time validation of an armstrong form. The form validation is defined in `inputValidators`, and
+ * the form state in `state`. This returns an object containing the following:
+ *  - `setTouched`: a method provided to tell when a form input control has been interacted with by the user. This must
+ * be used on each input with validation. As an example:
+ *
+ * ```
+ * <TextInput {...bind.text('city')} label="City" onBlur={setTouched} />
+ * ```
+ *
+ *  - `formIsValid`: a boolean flag to denote if the form is valid or invalid. This doesn't rely on user interaction, so
+ * is independant of error messages displayed on the form.
+ *  - `validationErrors`: an array populated with the errors to display on the form. Should be passed to the `DataForm` component:
+ *
+ * ```
+ * <DataForm validationResults={validationErrors} validationMode="both">
+ * ```
+ *
+ * @param inputValidators form validation
+ * @param dataBinder form state
+ */
+export function useValidatedForm<TData>(inputValidators: InputValidator[], state: TData, formProp: (...keyChain: KeyChain) => BindingTools<TData>) {
+  // Use this to calculate if this is the first render. If it is, then don't run the validation useEffect as it'll show all of the validation errors prematurely.
+  const firstRenderCheck = React.useRef(0);
+
+  // Run validation on individual inputs whenever they change value.
+  for (const inputValidator of inputValidators) {
+    React.useEffect(() => {
```

Comment:
Hooks inside loops is super dangerous, if the length of that array ever changes, React would error cus it keeps track of hooks by their length.

Could the loop not go inside the effect?

Reply:
The array can never change length, it's the inputValidators that are defined when passed to the hook. 
The thought process behind this was to make it more performant, i.e. each validator is only run when its specific input changes value. I can look into doing this outside of a loop though.